src/Orleans/Core/Grain.cs

```diff
@@ -332,5 +324,51 @@ protected virtual Task ReadStateAsync()
         {
             return storage.ReadStateAsync();
         }
+
+        void ILifecycleParticipant<GrainLifecycleStage>.Participate(ILifecycleObservable<GrainLifecycleStage> lifecycle)
+        {
+            lifecycle.Subscribe(GrainLifecycleStage.SetupState, this);
+        }
+
+        async Task ILifecycleObserver.OnStart(CancellationToken ct)
+        {
+            if (ct.IsCancellationRequested)
+                return;
+            IStorageProvider storageProvider = GetStorageProvider();
+            string grainTypeName = this.GetType().FullName;
```

Comment:
I think this is not backwards compatible. In the past we were using `$"{grainType}[{genericArguments}]` as the state name, where grain type is not even `type.FullName` or something like that.

Reply:
hmm.. looks like there was  bug..?  We load the initial sate using what you describe, but the storage bridge was setup with a different name, meaning the initial loaded state is different from the state the bridge uses.  Not sure how that got past tests, but that's a serious bug, unless I'm misunderstanding what I'm reading.

Reply:
Change was introduced in July of last year, so it may be lower impact than I'd expected.   https://github.com/dotnet/orleans/commit/f634bde01f8d8fac696c0086986a1bbb61815cd8
Anyone familiar with what this was for?  @ReubenBond? @jdom?

Reply:
Ah.  I see.  This addressed a problem which only existed because the bridge was not being used to read the initial state.  Current code should be fine.